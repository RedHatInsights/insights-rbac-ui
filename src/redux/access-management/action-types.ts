export const GET_PRINCIPAL_ACCESS = 'GET_PRINCIPAL_ACCESS' as const;
