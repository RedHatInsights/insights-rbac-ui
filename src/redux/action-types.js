export const FETCH_GROUP = 'FETCH_GROUP';
export const FETCH_SYSTEM_GROUP = 'FETCH_SYSTEM_GROUP';
export const FETCH_GROUPS = 'FETCH_GROUPS';

export const ADD_GROUP = 'ADD_GROUP';
export const UPDATE_GROUP = 'UPDATE_GROUP';
export const REMOVE_GROUPS = 'REMOVE_GROUPS';

export const FETCH_USERS = 'FETCH_USERS';

export const ADD_ROLE = 'ADD_ROLE';
export const FETCH_ROLE = 'FETCH_ROLE';
export const FETCH_ROLES = 'FETCH_ROLES';
export const REMOVE_ROLE = 'REMOVE_ROLE';
export const RESET_SELECTED_ROLE = 'RESET_SELECTED_ROLE';
export const FETCH_ROLES_FOR_WIZARD = 'FETCH_ROLES_FOR_WIZARD';

export const FETCH_GROUP_POLICIES = 'FETCH_GROUP_POLICIES';
export const FETCH_POLICY = 'FETCH_POLICY';
export const ADD_POLICY = 'ADD_POLICY';
export const UPDATE_POLICY = 'UPDATE_POLICY';
export const REMOVE_POLICY = 'REMOVE_POLICY';
export const RESET_SELECTED_GROUP = 'RESET_SELECTED_GROUP';

export const REMOVE_MEMBERS_FROM_GROUP = 'REMOVE_MEMBERS_FROM_GROUP';
export const ADD_MEMBERS_TO_GROUP = 'ADD_MEMBERS_TO_GROUP';

export const REMOVE_ROLES_FROM_GROUP = 'REMOVE_ROLES_FROM_GROUP';
export const ADD_ROLES_TO_GROUP = 'ADD_ROLES_TO_GROUP';
export const FETCH_ROLES_FOR_GROUP = 'FETCH_ROLES_FOR_GROUP';
export const FETCH_MEMBERS_FOR_GROUP = 'FETCH_MEMBERS_FOR_GROUP';
export const FETCH_ADD_ROLES_FOR_GROUP = 'FETCH_ADD_ROLES_FOR_GROUP';
export const FETCH_ROLE_FOR_USER = 'FETCH_ROLE_FOR_USER';
export const FETCH_ROLE_FOR_PRINCIPAL = 'FETCH_ROLE_FOR_PRINCIPAL';
export const UPDATE_ROLE = 'UPDATE_ROLE';

export const GET_PRINCIPAL_ACCESS = 'GET_PRINCIPAL_ACCESS';

export const LIST_PERMISSIONS = 'LIST_PERMISSIONS';
export const LIST_APPLICATION_OPTIONS = 'LIST_APPLICATION_OPTIONS';
export const LIST_RESOURCE_OPTIONS = 'LIST_RESOURCE_OPTIONS';
export const LIST_OPERATION_OPTIONS = 'LIST_OPERATION_OPTIONS';

export const GET_RESOURCE_DEFINITIONS = 'GET_RESOURCE_DEFINITIONS';
export const GET_RESOURCE = 'GET_RESOURCE';
