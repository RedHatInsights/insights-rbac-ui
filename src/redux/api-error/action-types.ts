export const API_ERROR = 'API_ERROR' as const;
