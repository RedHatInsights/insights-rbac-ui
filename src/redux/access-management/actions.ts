import { GET_PRINCIPAL_ACCESS } from './action-types';
import { GetPrincipalAccessParams, getPrincipalAccess as getPrincipalAccessHelper } from './helper';

// Using global ReduxAction interface from store.d.ts
export const getPrincipalAccess = (apiProps: GetPrincipalAccessParams): ReduxAction<Promise<unknown>> => ({
  type: GET_PRINCIPAL_ACCESS,
  payload: getPrincipalAccessHelper(apiProps),
});
