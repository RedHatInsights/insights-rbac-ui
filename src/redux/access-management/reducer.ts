import { GET_PRINCIPAL_ACCESS } from './action-types';
import { PaginationDefaultI, defaultSettings } from '../../helpers/pagination';

// Type definitions for access management
interface AccessItem {
  permission: string;
  application: string;
  resource_type: string;
  verb: string;
  resourceDefinitions?: unknown[];
  [key: string]: unknown;
}

interface AccessData {
  data: AccessItem[];
  meta: PaginationDefaultI;
}

export interface AccessState {
  isLoading: boolean;
  access: AccessData;
  [key: string]: unknown;
}

// Using global ReduxAction interface from store.d.ts instead of local Action interface

// Initial State
export const accessInitialState: AccessState = {
  isLoading: false,
  access: {
    data: [],
    meta: defaultSettings,
  },
};

const setLoadingState = (state: AccessState): AccessState => ({ ...state, isLoading: true });

const setAccess = (state: AccessState, action?: ReduxAction<AccessData>): AccessState => {
  if (!action) return state;
  const { payload } = action;
  return { ...state, access: payload, isLoading: false };
};

// Reducer function map with proper typing
type ReducerFunctions = Record<string, (state: AccessState, action?: ReduxAction<any>) => AccessState>;

const accessReducer: ReducerFunctions = {
  [`${GET_PRINCIPAL_ACCESS}_PENDING`]: setLoadingState,
  [`${GET_PRINCIPAL_ACCESS}_FULFILLED`]: setAccess,
};

export default accessReducer;
