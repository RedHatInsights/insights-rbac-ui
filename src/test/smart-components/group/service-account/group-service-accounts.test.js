import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import GroupServiceAccounts from '../../../../features/groups/service-account/group-service-accounts';
import { groupsInitialState } from '../../../../redux/groups/reducer';
import * as GroupActions from '../../../../redux/groups/actions';
import { FETCH_GROUP, FETCH_SERVICE_ACCOUNTS_FOR_GROUP, FETCH_SYSTEM_GROUP } from '../../../../redux/groups/action-types';

jest.mock('../../../../redux/groups/actions', () => {
  const actual = jest.requireActual('../../../../redux/groups/actions');
  return {
    __esModule: true,
    ...actual,
  };
});

describe('<GroupServiceAccounts />', () => {
  let initialProps;
  let mockStore;
  let initialState;

  const fetchGroupSpy = jest.spyOn(GroupActions, 'fetchGroup');
  const fetchSystemGroupSpy = jest.spyOn(GroupActions, 'fetchSystemGroup');
  const fetchServiceAccountsForGroupSpy = jest.spyOn(GroupActions, 'fetchServiceAccountsForGroup');

  beforeEach(() => {
    initialProps = {};
    mockStore = configureStore();
    initialState = {
      groupReducer: {
        ...groupsInitialState,
      },
    };
  });

  afterEach(() => {
    fetchGroupSpy.mockReset();
    fetchSystemGroupSpy.mockReset();
    fetchServiceAccountsForGroupSpy.mockReset();
  });

  it('should render correctly', () => {
    fetchGroupSpy.mockImplementationOnce(() => ({ type: FETCH_GROUP, payload: Promise.resolve({}) }));
    fetchSystemGroupSpy.mockImplementationOnce(() => ({ type: FETCH_SYSTEM_GROUP, payload: Promise.resolve({}) }));
    fetchServiceAccountsForGroupSpy.mockImplementationOnce(() => ({
      type: FETCH_SERVICE_ACCOUNTS_FOR_GROUP,
      payload: Promise.resolve({ data: [], meta: { count: 0 } }),
    }));
    const store = mockStore(initialState);
    const container = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/123/service-accounts']} initialIndex={0}>
          <Routes>
            <Route path="/groups/detail/:groupId/service-accounts" element={<GroupServiceAccounts />} {...initialProps} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );
    expect(screen.getByText('Add service account')).toBeVisible();
    expect(container.baseElement).toMatchSnapshot();
  });
});
