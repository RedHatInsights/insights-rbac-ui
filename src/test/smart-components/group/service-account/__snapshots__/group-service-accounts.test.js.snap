// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupServiceAccounts /> should render correctly 1`] = `
<body>
  <div>
    <section
      class="ins-l-content"
      id="tab-service-accounts"
    >
       
      <div
        class="pf-v5-c-alert pf-m-inline pf-m-plain pf-m-info rbac-service-accounts-alert"
        data-ouia-component-id="OUIA-Generated-Alert-info-1"
        data-ouia-component-type="PF5/Alert"
        data-ouia-safe="true"
      >
        <div
          class="pf-v5-c-alert__icon"
        >
          <svg
            aria-hidden="true"
            class="pf-v5-svg"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
            />
          </svg>
        </div>
        <h4
          class="pf-v5-c-alert__title"
        >
          <span
            class="pf-v5-screen-reader"
          >
            Info alert:
          </span>
          To add, reset credentials, or delete service accounts visit the {link}.
        </h4>
      </div>
      <div>
        <div
          class="pf-v5-c-toolbar  ins-c-primary-toolbar"
          data-ouia-component-id="PrimaryToolbar"
          data-ouia-component-type="PF5/Toolbar"
          data-ouia-safe="true"
          id="ins-primary-data-toolbar"
        >
          <div
            class="pf-v5-c-toolbar__content"
          >
            <div
              class="pf-v5-c-toolbar__content-section"
            >
              <div
                class="pf-v5-c-toolbar__group pf-m-filter-group ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg"
              >
                <div
                  class="pf-v5-c-toolbar__item"
                >
                  <button
                    aria-expanded="false"
                    class="pf-v5-c-menu-toggle"
                    data-ouia-component-id="OUIA-Generated-MenuToggle-1"
                    data-ouia-component-type="PF5/MenuToggle"
                    data-ouia-safe="true"
                    type="button"
                  >
                    <span
                      class="pf-v5-c-menu-toggle__text"
                    >
                      <label
                        class="pf-v5-c-check pf-m-standalone"
                      >
                        <input
                          aria-invalid="false"
                          aria-label="Select all"
                          class="pf-v5-c-check__input"
                          data-ouia-component-id="BulkSelectCheckbox"
                          data-ouia-component-type="PF5/MenuToggleCheckbox"
                          data-ouia-safe="true"
                          name="group-accounts-toggle-checkbox"
                          type="checkbox"
                        />
                      </label>
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__controls"
                    >
                      <span
                        class="pf-v5-c-menu-toggle__toggle-icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 320 512"
                          width="1em"
                        >
                          <path
                            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
                <div
                  class="pf-v5-c-toolbar__item ins-c-primary-toolbar__filter"
                >
                  <div
                    class="pf-v5-l-split ins-c-conditional-filter"
                  >
                    <div
                      class="pf-v5-l-split__item"
                    >
                      <button
                        aria-expanded="false"
                        aria-label="Conditional filter toggle"
                        class="pf-v5-c-menu-toggle ins-c-conditional-filter__group"
                        data-ouia-component-id="OUIA-Generated-MenuToggle-2"
                        data-ouia-component-type="PF5/MenuToggle"
                        data-ouia-safe="true"
                        type="button"
                      >
                        <span
                          class="pf-v5-c-menu-toggle__text"
                        >
                          <span
                            class="pf-v5-c-icon pf-m-md"
                          >
                            <span
                              class="pf-v5-c-icon__content"
                            >
                              <svg
                                aria-hidden="true"
                                class="pf-v5-svg"
                                fill="currentColor"
                                height="1em"
                                role="img"
                                viewBox="0 0 512 512"
                                width="1em"
                              >
                                <path
                                  d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
                                />
                              </svg>
                            </span>
                          </span>
                          <span
                            class="ins-c-conditional-filter__value-selector"
                          >
                            Client ID
                          </span>
                        </span>
                        <span
                          class="pf-v5-c-menu-toggle__controls"
                        >
                          <span
                            class="pf-v5-c-menu-toggle__toggle-icon"
                          >
                            <svg
                              aria-hidden="true"
                              class="pf-v5-svg"
                              fill="currentColor"
                              height="1em"
                              role="img"
                              viewBox="0 0 320 512"
                              width="1em"
                            >
                              <path
                                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                              />
                            </svg>
                          </span>
                        </span>
                      </button>
                    </div>
                    <div
                      class="pf-v5-l-split__item pf-m-fill"
                      data-ouia-component-id="ConditionalFilter"
                    >
                      <span
                        class="pf-v5-c-form-control pf-m-icon ins-c-conditional-filter"
                      >
                        <input
                          aria-invalid="false"
                          aria-label="text input"
                          data-ouia-component-id="ConditionalFilter"
                          data-ouia-component-type="PF5/TextInput"
                          data-ouia-safe="true"
                          id="filter-by-clientId"
                          placeholder="Filter by {key}"
                          type="text"
                          value=""
                          widget-type="InsightsInput"
                        />
                        <span
                          class="pf-v5-c-form-control__utilities"
                        >
                          <span
                            class="pf-v5-c-form-control__icon"
                          >
                            <span
                              class="pf-v5-c-icon pf-m-md ins-c-search-icon"
                            >
                              <span
                                class="pf-v5-c-icon__content"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="pf-v5-svg"
                                  fill="currentColor"
                                  height="1em"
                                  role="img"
                                  viewBox="0 0 512 512"
                                  width="1em"
                                >
                                  <path
                                    d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </span>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="pf-v5-c-toolbar__item ins-c-primary-toolbar__first-action pf-m-spacer-sm"
              >
                <span>
                  <a
                    class="rbac-m-hide-on-sm"
                    href="/iam/user-access/groups/detail/123/service-accounts/add-service-account"
                  >
                    <button
                      aria-disabled="false"
                      aria-label="Add service account to group"
                      class="pf-v5-c-button pf-m-primary rbac-m-hide-on-sm"
                      data-ouia-component-id="add-service-account-button"
                      data-ouia-component-type="PF5/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                      Add service account
                    </button>
                  </a>
                </span>
              </div>
              <div
                class="pf-v5-c-toolbar__item  ins-c-primary-toolbar__actions pf-m-spacer-sm"
              >
                <button
                  aria-expanded="false"
                  aria-label="kebab dropdown toggle"
                  class="pf-v5-c-menu-toggle pf-m-plain"
                  data-ouia-component-id="OUIA-Generated-MenuToggle-plain-1"
                  data-ouia-component-type="PF5/MenuToggle"
                  data-ouia-safe="true"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 192 512"
                    width="1em"
                  >
                    <path
                      d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="pf-v5-c-toolbar__item ins-c-primary-toolbar__pagination"
              >
                <div
                  class="pf-v5-c-pagination pf-m-compact"
                  data-ouia-component-id="CompactPagination"
                  data-ouia-component-type="PF5/Pagination"
                  data-ouia-safe="true"
                  id="options-menu-top-pagination"
                  style="--pf-v5-c-pagination__nav-page-select--c-form-control--width-chars: 2;"
                >
                  <div
                    class="pf-v5-c-pagination__total-items"
                  >
                    <b>
                      0
                       - 
                      0
                    </b>
                     
                    of
                     
                    <b>
                      0
                    </b>
                     
                  </div>
                  <button
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    class="pf-v5-c-menu-toggle pf-m-plain pf-m-text"
                    data-ouia-component-id="OUIA-Generated-MenuToggle-plainText-1"
                    data-ouia-component-type="PF5/MenuToggle"
                    data-ouia-safe="true"
                    id="options-menu-top-toggle"
                    type="button"
                  >
                    <span
                      class="pf-v5-c-menu-toggle__text"
                    >
                      <b>
                        0
                         - 
                        0
                      </b>
                       
                      of
                       
                      <b>
                        0
                      </b>
                       
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__controls"
                    >
                      <span
                        class="pf-v5-c-menu-toggle__toggle-icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 320 512"
                          width="1em"
                        >
                          <path
                            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                  <nav
                    aria-label="Pagination"
                    class="pf-v5-c-pagination__nav"
                  >
                    <div
                      class="pf-v5-c-pagination__nav-control"
                    >
                      <button
                        aria-disabled="true"
                        aria-label="Go to previous page"
                        class="pf-v5-c-button pf-m-plain pf-m-disabled"
                        data-action="previous"
                        data-ouia-component-id="OUIA-Generated-Button-plain-1"
                        data-ouia-component-type="PF5/Button"
                        data-ouia-safe="true"
                        disabled=""
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 256 512"
                          width="1em"
                        >
                          <path
                            d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                          />
                        </svg>
                      </button>
                    </div>
                    <div
                      class="pf-v5-c-pagination__nav-control"
                    >
                      <button
                        aria-disabled="true"
                        aria-label="Go to next page"
                        class="pf-v5-c-button pf-m-plain pf-m-disabled"
                        data-action="next"
                        data-ouia-component-id="OUIA-Generated-Button-plain-2"
                        data-ouia-component-type="PF5/Button"
                        data-ouia-safe="true"
                        disabled=""
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 256 512"
                          width="1em"
                        >
                          <path
                            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                          />
                        </svg>
                      </button>
                    </div>
                  </nav>
                </div>
              </div>
            </div>
          </div>
          <div
            class="pf-v5-c-toolbar__content pf-m-hidden"
            hidden=""
          >
            <div
              class="pf-v5-c-toolbar__group"
            />
          </div>
        </div>
        <table
          aria-label="service accounts table"
          class="ins-c-table-empty-state pf-v5-c-table pf-m-grid-md"
          data-ouia-component-id="group-accounts"
          data-ouia-component-type="PF5/Table"
          data-ouia-safe="true"
          role="grid"
        >
          <thead
            class="pf-v5-c-table__thead"
          >
            <tr
              class="pf-v5-c-table__tr"
              data-ouia-component-id="OUIA-Generated-TableRow-1"
              data-ouia-component-type="PF5/TableRow"
              data-ouia-safe="true"
            >
              <th
                class="pf-v5-c-table__th"
                data-key="0"
                data-label="Name"
                scope="col"
                tabindex="-1"
              >
                Name
              </th>
              <th
                class="pf-v5-c-table__th"
                data-key="1"
                data-label="Client ID"
                scope="col"
                tabindex="-1"
              >
                Client ID
              </th>
              <th
                class="pf-v5-c-table__th"
                data-key="2"
                data-label="Owner"
                scope="col"
                tabindex="-1"
              >
                Owner
              </th>
              <th
                class="pf-v5-c-table__th"
                data-key="3"
                data-label="Time created"
                scope="col"
                tabindex="-1"
              >
                Time created
              </th>
            </tr>
          </thead>
          <tbody
            class="pf-v5-c-table__tbody"
            role="rowgroup"
          >
            <tr
              class="pf-v5-c-table__tr"
              data-ouia-component-id="OUIA-Generated-TableRow-2"
              data-ouia-component-type="PF5/TableRow"
              data-ouia-safe="true"
            >
              <td
                class="pf-v5-c-table__td"
                colspan="4"
                data-key="0"
                data-label="Name"
                tabindex="-1"
              >
                <div
                  class="pf-v5-c-empty-state pf-m-sm"
                >
                  <div
                    class="pf-v5-c-empty-state__content"
                  >
                    <div
                      class="pf-v5-c-empty-state__header"
                    >
                      <div
                        class="pf-v5-c-empty-state__icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 512 512"
                          width="1em"
                        >
                          <path
                            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                          />
                        </svg>
                      </div>
                      <div
                        class="pf-v5-c-empty-state__title"
                      >
                        <h4
                          class="pf-v5-c-empty-state__title-text"
                        >
                          No matching {items} found
                        </h4>
                      </div>
                    </div>
                    <div
                      class="pf-v5-c-empty-state__body pf-v5-u-mb-md"
                    >
                      This filter criteria matches no {items}.
                       
                      <br />
                      Try changing your filter settings.
                       
                      <br />
                    </div>
                    <div
                      class="pf-v5-c-empty-state__footer"
                    >
                      <div
                        class="pf-v5-c-empty-state__actions"
                      >
                        <button
                          aria-disabled="false"
                          class="pf-v5-c-button pf-m-link"
                          data-ouia-component-id="clear-filters-button"
                          data-ouia-component-type="PF5/Button"
                          data-ouia-safe="true"
                          type="button"
                        >
                          Clear all filters
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
        <div
          class="pf-v5-c-toolbar ins-c-table__toolbar"
          data-ouia-component-id="OUIA-Generated-RHI/TableToolbar-true-1"
          data-ouia-component-type="RHI/TableToolbar"
          data-ouia-safe="true"
          id="pf-random-id-0"
        >
          <div
            class="pf-v5-c-pagination pf-m-bottom"
            data-ouia-component-id="OUIA-Generated-Pagination-bottom-1"
            data-ouia-component-type="PF5/Pagination"
            data-ouia-safe="true"
            id="options-menu-bottom-pagination"
            style="--pf-v5-c-pagination__nav-page-select--c-form-control--width-chars: 2;"
          >
            <button
              aria-expanded="false"
              aria-haspopup="listbox"
              class="pf-v5-c-menu-toggle pf-m-plain pf-m-text"
              data-ouia-component-id="OUIA-Generated-MenuToggle-plainText-2"
              data-ouia-component-type="PF5/MenuToggle"
              data-ouia-safe="true"
              id="options-menu-bottom-toggle"
              type="button"
            >
              <span
                class="pf-v5-c-menu-toggle__text"
              >
                <b>
                  0
                   - 
                  0
                </b>
                 
                of
                 
                <b>
                  0
                </b>
                 
              </span>
              <span
                class="pf-v5-c-menu-toggle__controls"
              >
                <span
                  class="pf-v5-c-menu-toggle__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </span>
            </button>
            <nav
              aria-label="Pagination"
              class="pf-v5-c-pagination__nav"
            >
              <div
                class="pf-v5-c-pagination__nav-control pf-m-first"
              >
                <button
                  aria-disabled="true"
                  aria-label="Go to first page"
                  class="pf-v5-c-button pf-m-plain pf-m-disabled"
                  data-action="first"
                  data-ouia-component-id="OUIA-Generated-Button-plain-3"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe="true"
                  disabled=""
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 448 512"
                    width="1em"
                  >
                    <path
                      d="M223.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L319.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L393.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34zm-192 34l136 136c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9L127.9 256l96.4-96.4c9.4-9.4 9.4-24.6 0-33.9L201.7 103c-9.4-9.4-24.6-9.4-33.9 0l-136 136c-9.5 9.4-9.5 24.6-.1 34z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="pf-v5-c-pagination__nav-control"
              >
                <button
                  aria-disabled="true"
                  aria-label="Go to previous page"
                  class="pf-v5-c-button pf-m-plain pf-m-disabled"
                  data-action="previous"
                  data-ouia-component-id="OUIA-Generated-Button-plain-4"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe="true"
                  disabled=""
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="pf-v5-c-pagination__nav-page-select"
              >
                <span
                  class="pf-v5-c-form-control pf-m-disabled"
                >
                  <input
                    aria-invalid="false"
                    aria-label="Current page"
                    data-ouia-component-id="OUIA-Generated-TextInputBase-2"
                    data-ouia-component-type="PF5/TextInput"
                    data-ouia-safe="true"
                    disabled=""
                    max="0"
                    min="1"
                    type="number"
                    value="0"
                  />
                </span>
                <span
                  aria-hidden="true"
                >
                  of
                   
                  0
                </span>
              </div>
              <div
                class="pf-v5-c-pagination__nav-control"
              >
                <button
                  aria-disabled="true"
                  aria-label="Go to next page"
                  class="pf-v5-c-button pf-m-plain pf-m-disabled"
                  data-action="next"
                  data-ouia-component-id="OUIA-Generated-Button-plain-5"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe="true"
                  disabled=""
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="pf-v5-c-pagination__nav-control pf-m-last"
              >
                <button
                  aria-disabled="true"
                  aria-label="Go to last page"
                  class="pf-v5-c-button pf-m-plain pf-m-disabled"
                  data-action="last"
                  data-ouia-component-id="OUIA-Generated-Button-plain-6"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe="true"
                  disabled=""
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 448 512"
                    width="1em"
                  >
                    <path
                      d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
                    />
                  </svg>
                </button>
              </div>
            </nav>
          </div>
          <div
            class="pf-v5-c-toolbar__content pf-m-hidden"
            hidden=""
          >
            <div
              class="pf-v5-c-toolbar__group"
            />
          </div>
        </div>
      </div>
       
    </section>
  </div>
</body>
`;
